### PR TITLE
Bugfix and improvement.

### PR DESCRIPTION
**Pull request by Magicloud**
- ASG user_data cannot be empty string. So rework default value.

- ASG does not take target group for ALB usage.

---
name: Pull request template
about: Make a PR to terraform-aws-foundation
---

Please include the following in your PR:

Please also note that these are not hard requirements, but merely serve to define
what maintainers are looking for in PR's.  Including these will more likely lead 
to your PR being reviewed and accepted.

- [ ] Update the changelog
- [ ] Make sure that modules and files are documented. This can be done inside the module and files.
- [ ] Make sure that new modules directories contain a basic README.md file.
- [ ] Make sure that the module is added to [tests/main.tf](https://github.com/fpco/terraform-aws-foundation/blob/master/tests/main.tf)
- [ ] Make sure that the linting passes on CI.
- [ ] Make sure that there is an up to date example for your code:
      - For new `modules` this would entail example code for how to use the module or some explanation in the module readme.
      - For new examples please provide a README explaining how to run the example. It's also ideal to provide a basic makefile to use the example as well.
- [ ] Make sure that there is a manual CI trigger that can test the deployment.

### CODE REVIEW

**Comment by Magicloud** [comment]
@qrilka I knew you used web-alb module. So please give advise here.

**Comment by qrilka** [comment]
Nah, in the end it didn't work out because of requirement of preexisting ACM certificate. So in the end I stole some code from https://gitlab.fpcomplete.com/greatcall/consumer-env/tree/master/terraform/modules/env-wrapper
I'll take a look into this tomorrow

File: examples/alb-test/tester.tf [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+}
+
+module "alb" {
+  source        = "../../modules/alb"
+  ports         = ["80"]
+  vpc_id        = module.vpc.vpc_id
+  sg_ids        = [aws_security_group.wide_open.id]
+  identifier    = "albest"
```
Comment (qrilka) [inline examples/alb-test/tester.tf]: does this identifier have any special meaning? I guess the first part is ALB but what does suffix mean? Shouldn't it be "alb_test" as above?

File: examples/alb-test/tester.tf [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  subnet_ids    = module.vpc.public_subnet_ids
+  ports_secured = []
+  dns_name      = "shida.dev-sandbox.fpcomplete.com"
```
Comment (qrilka) [inline examples/alb-test/tester.tf]: @Magicloud do we need to hardcode your name here and for the key name above?

File: examples/alb-test/tester.tf [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  acm_arn       = ""
+}
+
+data "aws_route53_zone" "dev-sandbox" {
+  name = "dev-sandbox.fpcomplete.com."
+}
+
+resource "aws_route53_record" "shida" {
+  zone_id = data.aws_route53_zone.dev-sandbox.zone_id
+  name    = "shida.${data.aws_route53_zone.dev-sandbox.name}"
```
Comment (qrilka) [inline examples/alb-test/tester.tf]: same here

File: examples/alb-test/tester.tf (hunk 1 above)
Comment (Magicloud) [reply examples/alb-test/tester.tf]: Ye.... There is a 6 char limit....

File: examples/alb-test/tester.tf (hunk 2 above)
Comment (Magicloud) [reply examples/alb-test/tester.tf]: It does not. I'd like this to be a unit test and planable without human interfere, so the code can be more robust.. But that would require a framework, that provides necessary variables in CI. Nothing done on that yet.

File: examples/alb-test/tester.tf (hunk 1 above)
Comment (qrilka) [reply examples/alb-test/tester.tf]: Maybe "albtst"? Sounds a bit more readable, BTW where does the limit come from?

File: examples/alb-test/tester.tf (hunk 2 above)
Comment (qrilka) [reply examples/alb-test/tester.tf]: If we don't have a good way resolve this then probably my comment could be ignored...

**Comment by ketzacoatl** [comment]
When using ALB's elsewhere, I generally found it best to keep the ALB separate from the Target Groups. Target Groups are more closely associated with ASGs (1:1). It's also questionable to combine the listeners with ALB.

File: examples/alb-test/tester.tf (hunk 1 above)
Comment (Magicloud) [reply examples/alb-test/tester.tf]: OK. For the limitation, I guess it is something missed in TF doc. For example, target group, says clearly that name_prefix has 6 char limitation. But for lb, it says nothing, but when planning, it throws error.

**Comment by Magicloud** [comment]
@ketzacoatl I have update the PR for some minor changes underneath. Please review.

**Comment by ketzacoatl** [comment]
@Magicloud can you confirm using this parameter, or using the ELB parameter, don't conflict with each other?

**Comment by Magicloud** [comment]
The PR seems overlap some new code in master. Closing.

**Pull request closed without merging**